metricbeat/module/system/cpu/metrics/metrics.go

```diff
@@ -0,0 +1,249 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package metrics
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/v7/libbeat/common"
+)
+
+// CPU manages the CPU metrics from /proc/stat
+// If a given metric isn't available on a given platform,
+// The value will be null. All methods that use these fields
+// should assume that any value can be null.
+// The values are in "ticks", which translates to milliseconds of CPU time
+type CPU struct {
+	user    *uint64 `metric:"user"`
+	sys     *uint64 `metric:"system"`
+	idle    *uint64 `metric:"idle"`
+	nice    *uint64 `metric:"nice"`    // Linux, Darwin, BSD
+	irq     *uint64 `metric:"irq"`     // Linux and openbsd
+	wait    *uint64 `metric:"iowait"`  // Linux and AIX
+	softIrq *uint64 `metric:"softirq"` // Linux only
+	stolen  *uint64 `metric:"steal"`   // Linux only
+}
+
+// CPUMetrics carries global and per-core CPU metrics
+type CPUMetrics struct {
+	totals CPU
+	// list carries the same data, broken down by CPU
+	list []CPU
+}
+
+// Total returns the total CPU time in ticks as scraped by the API
+func (cpu CPU) Total() uint64 {
+
+	var total uint64
+	fn := func(field reflect.Value, _ int, _ string) {
+		total = total + field.Uint()
+	}
+	cpu.iterateCPUWithFunc(fn)
+	return total
+
+}
+
+// iterateCPUWithFunc uses reflection to interate over the CPU struct, stopping at every non-null field
+// `field` is the value of the struct field, iter is the field's place in the struct, and name is the `metric` tag.
+func (cpu CPU) iterateCPUWithFunc(iterFunc func(field reflect.Value, iter int, name string)) {
```

Comment:
If possible I'd prefer to not work with reflect at all. I guess it is some workaround we will get rid of in the future?

Anyways, it is assumed that `field.Uint` returns a value and will not panic in functions that use the iterate helper. If we assume it is always uint (or can be converted to uint64), rather reflect that in the callback via: `func (cpu CPU) iterateCPUWithFunc(iterFunc func(name string, value uint64)) { ...`.

No need to pass the index, functions that require the index should be able to compute the index by themselves.

Reply:
You're right, that function signature makes a tad more sense.

As far as the reflection, it's mostly there so in service of limited the amount of duplicated code that ends up in the `metrics_[platform].go` files, since our way of distinguishing between "value does not exist" and "value is zero" in the structs is via pointers, which we then iterate over. Reflection isn't strictly necessary for that, but I'm not a huge fan of a huge list of `if struct.field ...` statements. I'm also not a huge fan of basing things off of hashmaps, which are more opaque. 

I've tried a few different ideas here, and landed on this one, since it's the least amount of duplicated code, and would require the least amount of tinkering if we decided to start changing the actual reported metrics, or their field names. That being said, I'm totally open to other suggestions.

Reply:
Using reflection might require some more care, especially if you want to reuse some functionality.
If you really want to use reflection, I'd say hide is as well as possible.
Right now the way reflection is used here, you require all fields in the CPU struct to be `*uint(x)`. We do not distinguish between optional and non-optional fields in the struct, because that would break the loop here. You can't add float, bool, enums, strings without introducing a panic.

Some fields reported via `/proc` are overloaded in its meaning, and act as indicator/flag if the value is one of `0-1` or `0-2`. In order to make sense of those you might require two fields in your struct. One bool field that indicates if the flag was set, and one optional metric if the flag is not set. Are there any string like fields we want to be able to report for other metricsets?

> but I'm not a huge fan of a huge list of if struct.field ... statements.

Right. Alternative to that is to introduce a helper:

```
reportOptUint := func(name string, value *uint) {
  if value != nil { iterFunc(name, *value) }
}

reportOptUint("user", cpu.user)
reportOptUint("sys", cpu.sys)
...
```

Or pass an interface that is more limited to what reflects support + provide a set of helper types like `OptUint`, which allows you to express your needs via types.

You have multiple options here, and I'm ok to keep it as is, but if you want to use reflection as shown here, also add a comment to the `cpu` type clarifing that all fields in the struct must be `*uint64`.

Reply:
> In order to make sense of those you might require two fields in your struct. One bool field that indicates if the flag was set, and one optional metric if the flag is not set. Are there any string like fields we want to be able to report for other metricsets?

So, to be clear, the code in this PR is fairly tied to the logic of reporting CPU metrics, which across platforms has fit just fine into the idiom of "list of uints." I'm still not 100% sure to what degree we want to duplicate anything other that the public APIs in `cpu/metrics` in other metricset refactors going forward. If we did want this code or logic to end up in something other than `cpu`, you'd be absolutely right, we should do it differently. 


> In order to make sense of those you might require two fields in your struct. One bool field that indicates if the flag was set, and one optional metric if the flag is not set. Are there any string like fields we want to be able to report for other metricsets?

This is something I was thinking about. Part of me is kind of tempted to duplicate the logic of Rust's [`Option<T>`](https://doc.rust-lang.org/std/option/) type, which could look something like this:

```
type cpu struct {
  Total OptionalUint
  System OptionalUint
....
}

type OptionalUint struct {
  Some bool
  Value uint64
}
```